Stripe/STPSourcePoller.m

```diff
@@ -0,0 +1,209 @@
+        [application endBackgroundTask:bgTaskID];
+        bgTaskID = UIBackgroundTaskInvalid;
+    }];
+    self.dataTask = [self.apiClient retrieveSourceWithId:self.sourceID
+                                            clientSecret:self.clientSecret
+                                      responseCompletion:^(STPSource *source, NSHTTPURLResponse *response, NSError *error) {
+                                          [self continueWithSource:source response:response error:error];
+                                          self.requestCount++;
+                                          self.dataTask = nil;
+                                          [application endBackgroundTask:bgTaskID];
+                                          bgTaskID = UIBackgroundTaskInvalid;
+                                      }];
+}
+
+- (void)continueWithSource:(STPSource *)source
+                  response:(NSHTTPURLResponse *)response
+                     error:(NSError *)error {
+    if (response) {
+        NSUInteger status = response.statusCode;
+        if (status >= 400 && status < 500) {
+            // Don't retry requests that 4xx
+            self.completion(self.latestSource, error);
+            [self stopPolling];
+        } else if (status == 200) {
+            self.pollInterval = DefaultPollInterval;
+            self.retryCount = 0;
+            // Only call completion if source.status has changed
+            if (!self.latestSource || source.status != self.latestSource.status) {
+                self.completion(source, nil);
+            }
+            self.latestSource = source;
+            if ([self shouldContinuePollingSource:source]) {
+                [self pollAfter:self.pollInterval lastError:nil];
+            } else {
+                [self stopPolling];
+            }
+        } else {
+            // Backoff and increment retry count
+            self.pollInterval = MIN(self.pollInterval*2, MaxPollInterval);
+            self.retryCount++;
+            [self pollAfter:self.pollInterval lastError:error];
+        }
+    } else {
+        // Retry if there's a connectivity error
+        if (error.code == kCFURLErrorNotConnectedToInternet ||
+            error.code == kCFURLErrorNetworkConnectionLost) {
+            self.retryCount++;
+            [self pollAfter:self.pollInterval lastError:error];
+        } else {
+            // Don't call completion if the request was cancelled
+            if (error.code != kCFURLErrorCancelled) {
+                self.completion(self.latestSource, error);
+            }
+            [self stopPolling];
+        }
+    }
+}
+
+- (BOOL)shouldContinuePollingSource:(nullable STPSource *)source {
+    if (!source) {
+        return NO;
+    }
+    return source.status == STPSourceStatusPending;
+}
+
+- (void)restartPolling {
+    if (self.pollingStopped) {
+        return;
+    }
+    self.pollingPaused = NO;
+    if (!self.timer && !self.dataTask) {
+        [self pollAfter:0 lastError:nil];
```

Comment:
Should we take into account the pollInterval here for backoff purposes? Eg if you get a failure in the background. Not sure this edge case is worth too much extra logic, but maybe?